controllers/storageoscluster/cli_operand.go

```diff
@@ -0,0 +1,151 @@
+package storageoscluster
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/ondat/operator-toolkit/declarative"
+	"github.com/ondat/operator-toolkit/declarative/kubectl"
+	"github.com/ondat/operator-toolkit/declarative/kustomize"
+	"github.com/ondat/operator-toolkit/declarative/transform"
+	eventv1 "github.com/ondat/operator-toolkit/event/v1"
+	"github.com/ondat/operator-toolkit/operator/v1/operand"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	kustomizetypes "sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/filesys"
+
+	storageoscomv1 "github.com/storageos/operator/api/v1"
+	"github.com/storageos/operator/internal/image"
+	stransform "github.com/storageos/operator/internal/transform"
+)
+
+const (
+	// cliPackage contains the resource manifests for cli operand.
+	cliPackage = "cli"
+
+	// Kustomize image name for container image.
+	kImageCLI = "cli"
+
+	// Related image environment variable.
+	cliImageEnvVar = "RELATED_IMAGE_CLI"
+)
+
+type CLIOperand struct {
+	name            string
+	client          client.Client
+	requires        []string
+	requeueStrategy operand.RequeueStrategy
+	fs              filesys.FileSystem
+	kubectlClient   kubectl.KubectlClient
+}
+
+var _ operand.Operand = &CLIOperand{}
+
+func (c *CLIOperand) Name() string                             { return c.name }
+func (c *CLIOperand) Requires() []string                       { return c.requires }
+func (c *CLIOperand) RequeueStrategy() operand.RequeueStrategy { return c.requeueStrategy }
+func (c *CLIOperand) ReadyCheck(ctx context.Context, obj client.Object) (bool, error) {
+	ctx, span, _, log := instrumentation.Start(ctx, "CLIOperand.ReadyCheck")
+	defer span.End()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
+
+	// Get the deployment object and check status of the replicas. One ready
+	// replica should be enough for the installation to continue.
+	cliDep := &appsv1.Deployment{}
+	key := client.ObjectKey{Name: "storageos-cli", Namespace: obj.GetNamespace()}
+	if err := c.client.Get(ctx, key, cliDep); err != nil {
+		return false, err
+	}
+
+	if cliDep.Status.AvailableReplicas > 0 {
+		log.Info("Found available replicas more than 0", "availableReplicas", cliDep.Status.AvailableReplicas)
+		return true, nil
+	}
+
+	log.Info("storageos-cli not ready")
+	return false, nil
+}
+func (c *CLIOperand) PostReady(ctx context.Context, obj client.Object) error { return nil }
+
+func (c *CLIOperand) Ensure(ctx context.Context, obj client.Object, ownerRef metav1.OwnerReference) (eventv1.ReconcilerEvent, error) {
```

Comment:
As i understand it deploys CLI every time. Does it makes sense to make it configurable?

Reply:
I was thinking about this - We will be forwarding plugin requests to the CLI pod like `get volumes` etc. So the user could end up in a situation where they are trying to run these commands, but they have no CLI pod deployed to process the request. For that reason I think it's best to deploy the CLI by default - but we could have a cluster spec field to disable it if you think that's necessary or useful? We do have precedent for this with the `disableScheduler` field

Reply:
I think it still makes sense to install the cli deployment via the operator as go-cli and the operator have the same release cycle and it is convenient to update the related image for a new release. If we go the plugin route we will need to discover the latest go-cli release via the github API, which is fine for the most part, but adds another flag to set when doing airgap installs.

Reply:
There are other options as well, for example, harmonize operator<->cli versions and use the same as operator. Or maintain some metadata at the plugin and update it when we release a new version of CLI tool (if they have a different release cycle).

Grabbing the latest isn't the optimal solution because we have to guarantee cli works ideally with node.

The other reason i think a temporary deployment makes more sense is CLI is a fat one full of other binaries and root user. So i think if we would like to deploy as part of the product we have to increase the security of it. WDYT?

Reply:
When you say "temporary deployment" do you mean spin up the deployment just to execute a command and then take it down again? (like we do for the etcd validation). If so, I think it might be very slow and a poor UX

Reply:
I did have this thought originally, but it seems like a lot of overhead (spinning up the pod, cleaning up after.. then there are situations where it fails, or where the user executes multiple commands in quick succession and we get multiple cli pods and potential complications there etc) - it just feels like there are a lot of edge cases and potentials for it to cause a user havoc and this solution was supposed to be a "quick-win" 🙃.

1 second+ is relatively quick, but not when you expect instant feedback like every other `kubectl` command - people are impatient! 😁  

So IMO the CLI deployment needs to stay in running phase, but if you feel strongly otherwise then we can discuss offline and get some more opinions?

Reply:
Ok makes sense.